### PR TITLE
Remove replica row when shutting down the MySQL heartbeater.

### LINKED ISSUE #74: Immediately delete replica row when mysql heartbeater exits.
Removing the row immediately will allow the discoverer to poll with a smaller delay, decreasing the amount of time that the now-offline replica spends in the partitioning hash ring.

### PR DESCRIPTION
**Pull request by jmalloc**
Fixes #74.

I implemented this by adding a `Shutdown()` method to service which allows for
a time limited "graceful shutdown" period.

### CODE REVIEW

File: service/behavior.go [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package service
+
+import "context"
+
+// NoTimeoutBehavior can be embedded in Service implementations to indicate that
```
Comment (koden-km) [inline service/behavior.go]:
```suggestion
// NoShutdownBehavior can be embedded in Service implementations to indicate that
```

File: service/many_test.go [hunk 2]
```diff
@@ -54,29 +55,35 @@ var _ = Describe("func Run()", func() {
...
+			1*time.Second,
 			[]Service{s1, s2},
 			observer.Ignore,
 		)
 		Expect(err).ShouldNot(HaveOccurred())
 	})
 
-	It("records facts when a service starts and stops", func() {
-		expect := errors.New("<error>")
+	It("records facts when a service starts and stops and shutsdown", func() {
```
Comment (koden-km) [inline service/many_test.go]: _[minor note]_ There are a lot of "and"'s. Should this be worded this better with a comma?
> starts and stops and shutsdown

File: service/service.go [hunk 3]
```diff
@@ -13,8 +13,21 @@ type Service interface {
 	// "mysql/replica.presence.heartbeater".
 	Name() string
 
-	// It is called when the engine is started. ctx is canceled when the engine
-	// is stopped for any reason. If it returns an error before ctx is canceled,
-	// the engine is stopped.
+	// Run is called when the engine is started. ctx is canceled when the engine
+	// is stopped for any reason.
+	//
+	// It if returns nil before ctx is canceled, it Shutdown() is called
```
Comment (koden-km) [inline service/service.go]: Something doesn't sound right here?
> before ctx is canceled, **it** Shutdown() is called

File: service/many_test.go (hunk 2 above)
Comment (jmalloc) [reply service/many_test.go]: Yeah, should probs be `starts, stops and shutsdown`, ta.

File: service/service.go (hunk 3 above)
Comment (jmalloc) [reply service/service.go]: Will check, I think the word `it` is just superfluous.

File: service/many_test.go (hunk 2 above)
Comment (danilvpetrov) [reply service/many_test.go]: shouldn't it be `shuts down` (spelled separately)?

Comment (jmalloc) [reply service/many_test.go]: Yeah, that's better too.

**Comment by jmalloc** [comment]
All updated.

**Pull request merged**